In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn import linear_model

In [2]:
import operator
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import PolynomialFeatures

In [3]:
import seaborn as sns

In [4]:
df = pd.read_csv('https://raw.githubusercontent.com/maxtran1997gmail/Files/main/dataset_2190_cholesterol.csv')
df.head(10)


age  sex  cp  trestbps  fbs  restecg  thalach  exang  oldpeak  slope ca  \
0   63    1   1       145    1        2      150      0      2.3      3  0   
1   67    1   4       160    0        2      108      1      1.5      2  3   
2   67    1   4       120    0        2      129      1      2.6      2  2   
3   37    1   3       130    0        0      187      0      3.5      3  0   
4   41    0   2       130    0        2      172      0      1.4      1  0   
5   56    1   2       120    0        0      178      0      0.8      1  0   
6   62    0   4       140    0        2      160      0      3.6      3  2   
7   57    0   4       120    0        0      163      1      0.6      1  0   
8   63    1   4       130    0        2      147      0      1.4      2  1   
9   53    1   4       140    1        2      155      1      3.1      3  0   

  thal  num  chol  
0    6    0   233  
1    3    2   286  
2    7    1   229  
3    3    0   250  
4    3    0   204  
5    3    0   236  
6    3    3   268  
7    3    0   354  
8    7    2   254  
9    7    1   203

1. sex : 1 = male; 0 = female
2. cp: value 1: typical angina; value 2: atypical angina; value 3:non-anginal pain; value 4: asymptomatic
3. trestbps : resting blood pressure (in mm Hg on admission to the hospital)
4. fbs: fasting blood sugar > 120 mg/dl; (1 = true; 0 = false)
5. restecg: resting electrocardiographic results 
-- Value 0: normal 
-- Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV) 
-- Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria 20 ekgmo (month of exercise ECG reading)
6. thalach: maximum heart rate achieved
7. exang: exercise induced angina (1 = yes; 0 = no) 
8. oldpeak: ST depression induced by exercise relative to rest
9. slope: the slope of the peak exercise ST segment 
-- Value 1: upsloping 
-- Value 2: flat 
-- Value 3: downsloping
9. ca: number of major vessels (0-3) colored by flourosopy
10. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect 
11. num: diagnosis of heart disease (angiographic disease status) 
-- Value 0: < 50% diameter narrowing 
-- Value 1: > 50% diameter narrowing (in any major vessel: attributes 59 through 68 are vessels)
12. chol: serum cholestoral in mg/dl 

In [5]:
df.describe(include="all")

age         sex          cp    trestbps         fbs  \
count   303.000000  303.000000  303.000000  303.000000  303.000000   
unique         NaN         NaN         NaN         NaN         NaN   
top            NaN         NaN         NaN         NaN         NaN   
freq           NaN         NaN         NaN         NaN         NaN   
mean     54.438944    0.679868    3.158416  131.689769    0.148515   
std       9.038662    0.467299    0.960126   17.599748    0.356198   
min      29.000000    0.000000    1.000000   94.000000    0.000000   
25%      48.000000    0.000000    3.000000  120.000000    0.000000   
50%      56.000000    1.000000    3.000000  130.000000    0.000000   
75%      61.000000    1.000000    4.000000  140.000000    0.000000   
max      77.000000    1.000000    4.000000  200.000000    1.000000   

           restecg     thalach       exang     oldpeak       slope   ca thal  \
count   303.000000  303.000000  303.000000  303.000000  303.000000  303  303   
unique         NaN         NaN         NaN         NaN         NaN    5    4   
top            NaN         NaN         NaN         NaN         NaN    0    3   
freq           NaN         NaN         NaN         NaN         NaN  176  166   
mean      0.990099  149.607261    0.326733    1.039604    1.600660  NaN  NaN   
std       0.994971   22.875003    0.469794    1.161075    0.616226  NaN  NaN   
min       0.000000   71.000000    0.000000    0.000000    1.000000  NaN  NaN   
25%       0.000000  133.500000    0.000000    0.000000    1.000000  NaN  NaN   
50%       1.000000  153.000000    0.000000    0.800000    2.000000  NaN  NaN   
75%       2.000000  166.000000    1.000000    1.600000    2.000000  NaN  NaN   
max       2.000000  202.000000    1.000000    6.200000    3.000000  NaN  NaN   

               num        chol  
count   303.000000  303.000000  
unique         NaN         NaN  
top            NaN         NaN  
freq           NaN         NaN  
mean      0.937294  246.693069  
std       1.228536   51.776918  
min       0.000000  126.000000  
25%       0.000000  211.000000  
50%       0.000000  241.000000  
75%       2.000000  275.000000  
max       4.000000  564.000000

In [444]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   fbs       303 non-null    int64  
 5   restecg   303 non-null    int64  
 6   thalach   303 non-null    int64  
 7   exang     303 non-null    int64  
 8   oldpeak   303 non-null    float64
 9   slope     303 non-null    int64  
 10  ca        303 non-null    object 
 11  thal      303 non-null    object 
 12  num       303 non-null    int64  
 13  chol      303 non-null    int64  
dtypes: float64(1), int64(11), object(2)
memory usage: 33.3+ KB


In [445]:
df.isna().sum()

age         0
sex         0
cp          0
trestbps    0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
chol        0
dtype: int64

In [6]:
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder

In [447]:
#le = preprocessing.LabelEncoder()
#df = df.apply(LabelEncoder().fit_transform)
#df.head(10)

In [7]:
df.corr()

age       sex        cp  trestbps       fbs   restecg  \
age       1.000000 -0.097542  0.104139  0.284946  0.118530  0.148868   
sex      -0.097542  1.000000  0.010084 -0.064456  0.047862  0.021647   
cp        0.104139  0.010084  1.000000 -0.036077 -0.039975  0.067505   
trestbps  0.284946 -0.064456 -0.036077  1.000000  0.175340  0.146560   
fbs       0.118530  0.047862 -0.039975  0.175340  1.000000  0.069564   
restecg   0.148868  0.021647  0.067505  0.146560  0.069564  1.000000   
thalach  -0.393806 -0.048663 -0.334422 -0.045351 -0.007854 -0.083389   
exang     0.091661  0.146201  0.384060  0.064762  0.025665  0.084867   
oldpeak   0.203805  0.102173  0.202277  0.189171  0.005747  0.114133   
slope     0.161770  0.037533  0.152050  0.117382  0.059894  0.133946   
num       0.222853  0.224469  0.407075  0.157754  0.059186  0.183696   
chol      0.208950 -0.199915  0.072319  0.130120  0.009841  0.171043   

           thalach     exang   oldpeak     slope       num      chol  
age      -0.393806  0.091661  0.203805  0.161770  0.222853  0.208950  
sex      -0.048663  0.146201  0.102173  0.037533  0.224469 -0.199915  
cp       -0.334422  0.384060  0.202277  0.152050  0.407075  0.072319  
trestbps -0.045351  0.064762  0.189171  0.117382  0.157754  0.130120  
fbs      -0.007854  0.025665  0.005747  0.059894  0.059186  0.009841  
restecg  -0.083389  0.084867  0.114133  0.133946  0.183696  0.171043  
thalach   1.000000 -0.378103 -0.343085 -0.385601 -0.415040 -0.003432  
exang    -0.378103  1.000000  0.288223  0.257748  0.397057  0.061310  
oldpeak  -0.343085  0.288223  1.000000  0.577537  0.504092  0.046564  
slope    -0.385601  0.257748  0.577537  1.000000  0.377957 -0.004062  
num      -0.415040  0.397057  0.504092  0.377957  1.000000  0.070909  
chol     -0.003432  0.061310  0.046564 -0.004062  0.070909  1.000000

In [449]:
temp = df.corr()['chol']
temp[abs(temp >= 0.1)]

age         0.208950
trestbps    0.130120
restecg     0.171043
chol        1.000000
Name: chol, dtype: float64

In [8]:
feature_cols = ["age","trestbps","restecg"]
x = df[feature_cols].values
y = df["chol"].values

In [10]:
from sklearn import preprocessing
scaler = preprocessing.StandardScaler().fit(x)
scaler

StandardScaler()

In [451]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(x,y,test_size=0.20,random_state=0)

In [452]:
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

LinearRegression()

In [461]:
lin_reg.coef_

array([0.80090476, 0.28624381, 7.82997038])

In [453]:
lin_reg.score(x,y)

0.0641318623819166

coefficient of determination : 0.0641318623819166

In [459]:
X_test = np.array([62, 140, 2]).reshape(1, -1)

In [460]:
y_pred = lin_reg.predict(X_test)
y_pred

array([261.2995283])

Chol dự đoán = 261.29 khi age = 62; trestbps = 140; restecg = 2

In [456]:
data_mean= df.mean()
data_std = df.std()

cut_off = data_std
lower, upper = data_mean - cut_off, data_mean + cut_off

C:\Users\84776\AppData\Local\Temp\ipykernel_21380\2846444933.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  data_mean= df.mean()
C:\Users\84776\AppData\Local\Temp\ipykernel_21380\2846444933.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  data_std = df.std()


In [457]:
from sklearn.neighbors import LocalOutlierFactor
from sklearn.metrics import mean_absolute_error

In [458]:
lof = LocalOutlierFactor()
yhat = lof.fit_predict(X_train)
# select all rows that are not outliers
mask = yhat != -1
X_train, y_train = X_train[mask, :], y_train[mask]
# summarize the shape of the updated training dataset
print(X_train.shape, y_train.shape)
# fit the model
df = LinearRegression()
df.fit(X_train, y_train)
# evaluate the model
yhat = df.predict(X_test)
# evaluate predictions
mae = mean_absolute_error(y_test, yhat)
print('MAE: %.3f' % mae)

(234, 3) (234,)


ValueError: Found input variables with inconsistent numbers of samples: [61, 1]

In [ ]:
age_group = np.linspace(min(df["age"]),max(df["age"]),4)
age_group
age_names = ['Young','Mid-age','Old']

In [ ]:
age_group

In [ ]:
df['age_groups'] = pd.cut(df['age'], age_group, labels = age_names, include_lowest= True)
df[['age',"age_groups"]].tail(10)

In [ ]:
df["age_groups"].value_counts()

In [ ]:
plt.hist(df["trestbps"])
plt.xlabel("trestbps")
plt.ylabel("count")

In [ ]:
sns.regplot(x="age", y="chol", data=df)